File: app/src/main/java/com/ramez/shopp/Activities/InvoiceInfoActivity.java
package com.ramez.shopp.Activities;

import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import androidx.recyclerview.widget.LinearLayoutManager;

import com.ramez.shopp.Adapter.OrderProductsAdapter;
import com.ramez.shopp.ApiHandler.DataFeacher;
import com.ramez.shopp.Classes.Constants;
import com.ramez.shopp.Classes.MessageEvent;
import com.ramez.shopp.Classes.UtilityApp;
import com.ramez.shopp.Models.ItemDetailsModel;
import com.ramez.shopp.Models.OrderItemDetail;
import com.ramez.shopp.Models.OrderNewModel;
import com.ramez.shopp.Models.ResultAPIModel;
import com.ramez.shopp.R;
import com.ramez.shopp.Utils.NumberHandler;
import com.ramez.shopp.databinding.ActivityInvoiceInfoBinding;

import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import es.dmoral.toasty.Toasty;

public class InvoiceInfoActivity extends ActivityBase {
    public String currency = "BHD";
    ActivityInvoiceInfoBinding binding;
    //    List<OrderItemDetail> list;
    ItemDetailsModel orderModel;
    private OrderProductsAdapter orderProductsAdapter;
    private int orderId = 0;
    int store_id, userId;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityInvoiceInfoBinding.inflate(getLayoutInflater());
        View view = binding.getRoot();
        setContentView(view);

//        list = new ArrayList<>();
        currency = UtilityApp.getLocalData().getCurrencyCode();
        if (UtilityApp.isLogin()) {
            userId = UtilityApp.getUserData().getId();
        }

        if (UtilityApp.getLocalData() != null) {
            store_id = Integer.parseInt(UtilityApp.getLocalData().getCityId());
        }

        getIntentData();

        setTitle(R.string.invoice_details);

        binding.reOrderBut.setOnClickListener(view1 -> {
            for (int i = 0; i < orderModel.getOrderItemDetails().size(); i++) {
                OrderItemDetail orderProductsDM = orderModel.getOrderItemDetails().get(i);

                int count = orderProductsDM.getQuantity();
                int userId = UtilityApp.getUserData().getId();
                int storeId = Integer.parseInt(UtilityApp.getLocalData().getCityId());
                int productId = orderProductsDM.getProductId();
                int product_barcode_id = orderProductsDM.getProductBarcodeId();
                addToCart(view1, i, productId, product_barcode_id, count, userId, storeId);


            }

        });

    }

    private void getIntentData() {
        Bundle bundle = getIntent().getExtras();
        if (bundle != null) {
            OrderNewModel orderNewModel = (OrderNewModel) bundle.getSerializable(Constants.ORDER_MODEL);
            orderId = orderNewModel.getId();
            //  list = orderModel.getOrderProductsDMS();
            getProductList(orderId, userId, store_id, Constants.user_type);
        }
    }

    public void initAdapter() {
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getActiviy());
        binding.productsRecycler.setLayoutManager(linearLayoutManager);

        orderProductsAdapter = new OrderProductsAdapter(getActiviy(), orderModel.getOrderItemDetails());
        binding.productsRecycler.setAdapter(orderProductsAdapter);
        if (orderModel.getTotalAmount() > 0) {
            binding.tvTotalPrice.setText((NumberHandler.formatDouble(orderModel.getTotalAmount(), UtilityApp.getLocalData().getFractional()) + " " + currency));

        }


    }

    private void addToCart(View v, int i, int productId, int product_barcode_id, int quantity, int userId, int storeId) {
        new DataFeacher(false, (obj, func, IsSuccess) -> {

            if (IsSuccess) {

                UtilityApp.updateCart(1, orderModel.getOrderItemDetails().size());


                if (i == orderModel.getOrderItemDetails().size() - 1) {
                    initSnackBar(" " + getResources().getString(R.string.success_to_update_cart), v);

                }


            } else {

                initSnackBar(getString(R.string.fail_to_update_cart), v);
            }


        }).addCartHandle(productId, product_barcode_id, quantity, userId, storeId);
    }


    private void initSnackBar(String message, View viewBar) {
        Toasty.success(getActiviy(), message, Toast.LENGTH_SHORT, true).show();

    }


    public void getProductList(int order_id, int user_id, int store_id, String type) {

//        orderModel.getOrderItemDetails().clear();

        binding.loadingProgressLY.loadingProgressLY.setVisibility(View.VISIBLE);
        binding.dataLY.setVisibility(View.GONE);
        binding.noDataLY.noDataLY.setVisibility(View.GONE);
        binding.failGetDataLY.failGetDataLY.setVisibility(View.GONE);

        new DataFeacher(false, (obj, func, IsSuccess) -> {
            ResultAPIModel<ItemDetailsModel> result = (ResultAPIModel<ItemDetailsModel>) obj;

            String message = getString(R.string.fail_to_get_data);

            binding.loadingProgressLY.loadingProgressLY.setVisibility(View.GONE);

            if (func.equals(Constants.ERROR)) {

                if (result != null && result.message != null) {
                    message = result.message;
                }
                binding.dataLY.setVisibility(View.GONE);
                binding.noDataLY.noDataLY.setVisibility(View.GONE);
                binding.failGetDataLY.failGetDataLY.setVisibility(View.VISIBLE);
                binding.failGetDataLY.failTxt.setText(message);

            } else if (func.equals(Constants.FAIL)) {

                binding.dataLY.setVisibility(View.GONE);
                binding.noDataLY.noDataLY.setVisibility(View.GONE);
                binding.failGetDataLY.failGetDataLY.setVisibility(View.VISIBLE);
                binding.failGetDataLY.failTxt.setText(message);


            } else if (func.equals(Constants.NO_CONNECTION)) {
                binding.failGetDataLY.failGetDataLY.setVisibility(View.VISIBLE);
                binding.failGetDataLY.failTxt.setText(R.string.no_internet_connection);
                binding.failGetDataLY.noInternetIv.setVisibility(View.VISIBLE);
                binding.dataLY.setVisibility(View.GONE);

            } else {
                if (IsSuccess) {
                    if (result.data != null &&result.data.getOrderItemDetails() != null&& result.data.getOrderItemDetails().size() > 0) {

                        binding.dataLY.setVisibility(View.VISIBLE);
                        binding.noDataLY.noDataLY.setVisibility(View.GONE);
                        binding.failGetDataLY.failGetDataLY.setVisibility(View.GONE);

                        orderModel = result.data;
//                        list = result.data.getOrderItemDetails();

                        initAdapter();


                    } else {

                        binding.dataLY.setVisibility(View.GONE);
                        binding.noDataLY.noDataLY.setVisibility(View.VISIBLE);

                    }


                } else {

                    binding.dataLY.setVisibility(View.GONE);
                    binding.noDataLY.noDataLY.setVisibility(View.GONE);
                    binding.failGetDataLY.failGetDataLY.setVisibility(View.VISIBLE);
                    binding.failGetDataLY.failTxt.setText(message);


                }
            }

        }).getOrderDetails(order_id, user_id, store_id, type);
    }

}

File: app/src/main/java/com/ramez/shopp/activities/InvoiceInfoActivity.kt
package com.ramez.shopp.activities


import android.annotation.SuppressLint
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import com.ramez.shopp.ApiHandler.DataFeacher
import com.ramez.shopp.ApiHandler.DataFetcherCallBack
import com.ramez.shopp.Classes.Constants
import com.ramez.shopp.Classes.UtilityApp
import com.ramez.shopp.Models.*
import com.ramez.shopp.R
import com.ramez.shopp.Utils.NumberHandler
import com.ramez.shopp.adapter.OrderProductsAdapter
import com.ramez.shopp.databinding.ActivityInvoiceInfoBinding
import es.dmoral.toasty.Toasty


class InvoiceInfoActivity : ActivityBase() {
    var currency = "BHD"
    lateinit var binding: ActivityInvoiceInfoBinding

    //    List<OrderItemDetail> list;
    var orderModel: ItemDetailsModel? = null
    private var orderProductsAdapter: OrderProductsAdapter? = null
    private var orderId = 0
    var storeId1 = 0
    var userId = 0
    var fraction = 2
    var localModel: LocalModel? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityInvoiceInfoBinding.inflate(layoutInflater)
        val view: View = binding.root
        setContentView(view)

//        list = new ArrayList<>();
        localModel =
            if (UtilityApp.getLocalData() != null) UtilityApp.getLocalData() else UtilityApp.getDefaultLocalData(
                activity
            )
        currency = localModel?.currencyCode ?:Constants.CURRENCY
        fraction = localModel?.fractional ?:Constants.three
        if (UtilityApp.isLogin()) {
            userId = UtilityApp.getUserData().id
        }
        storeId1 = localModel?.cityId?.toInt() ?: Constants.default_storeId.toInt()
        intentData
        setTitle(R.string.invoice_details)
        binding.reOrderBut.setOnClickListener { view1 ->
            orderModel?.orderItemDetails?.indices?.forEach { i ->
                val orderProductsDM = orderModel?.orderItemDetails!![i]
                val count = orderProductsDM.quantity
                val userId = UtilityApp.getUserData().id
                val storeId = localModel?.cityId?.toInt()?:Constants.default_storeId.toInt()
                val productId = orderProductsDM.productId
                val productBarcodeId = orderProductsDM.productBarcodeId
                addToCart(view1, i, productId, productBarcodeId, count, userId, storeId)
            }
        }
    }

    //  list = orderModel.getOrderProductsDMS();
    private val intentData: Unit
        get() {
            val bundle = intent.extras
            if (bundle != null) {
                val orderNewModel = bundle.getSerializable(Constants.ORDER_MODEL) as OrderNewModel?
                orderId = orderNewModel?.id?:0
                //  list = orderModel.getOrderProductsDMS();
                getProductList(orderId, userId, storeId1, Constants.user_type)
            }
        }

    @SuppressLint("SetTextI18n")
    fun initAdapter() {
        val linearLayoutManager = LinearLayoutManager(activity)
        binding.productsRecycler.layoutManager = linearLayoutManager
        orderProductsAdapter = OrderProductsAdapter(activity, orderModel?.orderItemDetails)
        binding.productsRecycler.adapter = orderProductsAdapter
        if (orderModel?.totalAmount?:0.0 > 0) {
            binding.tvTotalPrice.text =
                NumberHandler.formatDouble(orderModel?.totalAmount, fraction) + " " + currency
        }
    }

    private fun addToCart(
        v: View,
        i: Int,
        productId: Int,
        product_barcode_id: Int,
        quantity: Int,
        userId: Int,
        storeId: Int
    ) {
        if (quantity > 0) {
            DataFeacher(false,object :
                DataFetcherCallBack {
                override fun Result(obj: Any?, func: String?, IsSuccess: Boolean) {
                    if (IsSuccess) {
                        UtilityApp.updateCart(1, orderModel?.orderItemDetails?.size?:0)
                        if (i == (orderModel?.orderItemDetails?.size ?: 0) - 1) {
                            initSnackBar(" " + resources.getString(R.string.success_to_update_cart), v)
                        }
                    } else {
                        initSnackBar(getString(R.string.fail_to_update_cart), v)
                    }                }


            }).addCartHandle(productId, product_barcode_id, quantity, userId, storeId)
        } else {
            Toast(getString(R.string.quanity_wrong))
        }
    }

    private fun initSnackBar(message: String, viewBar: View) {
        Toasty.success(activity, message, Toast.LENGTH_SHORT, true).show()
    }

    fun getProductList(order_id: Int, user_id: Int, store_id: Int, type: String?) {

//        orderModel.getOrderItemDetails().clear();
        binding.loadingProgressLY.loadingProgressLY.visibility = View.VISIBLE
        binding.dataLY.visibility = View.GONE
        binding.noDataLY.noDataLY.visibility = View.GONE
        binding.failGetDataLY.failGetDataLY.visibility = View.GONE
        DataFeacher(false, object :DataFetcherCallBack {
            override fun Result(obj: Any?, func: String?, IsSuccess: Boolean) {
                val result = obj as ResultAPIModel<ItemDetailsModel?>?
                var message: String? = getString(R.string.fail_to_get_data)
                binding.loadingProgressLY.loadingProgressLY.visibility = View.GONE
                if (func == Constants.ERROR) {
                    if (result?.message != null) {
                        message = result.message
                    }
                    binding.dataLY.visibility = View.GONE
                    binding.noDataLY.noDataLY.visibility = View.GONE
                    binding.failGetDataLY.failGetDataLY.visibility = View.VISIBLE
                    binding.failGetDataLY.failTxt.text = message
                } else if (func == Constants.FAIL) {
                    binding.dataLY.visibility = View.GONE
                    binding.noDataLY.noDataLY.visibility = View.GONE
                    binding.failGetDataLY.failGetDataLY.visibility = View.VISIBLE
                    binding.failGetDataLY.failTxt.text = message
                } else if (func == Constants.NO_CONNECTION) {
                    binding.failGetDataLY.failGetDataLY.visibility = View.VISIBLE
                    binding.failGetDataLY.failTxt.setText(R.string.no_internet_connection)
                    binding.failGetDataLY.noInternetIv.visibility = View.VISIBLE
                    binding.dataLY.visibility = View.GONE
                } else {
                    if (IsSuccess) {
                        if (result?.data != null && result.data?.orderItemDetails != null && result.data?.orderItemDetails?.size?:0 > 0) {
                            binding.dataLY.visibility = View.VISIBLE
                            binding.noDataLY.noDataLY.visibility = View.GONE
                            binding.failGetDataLY.failGetDataLY.visibility = View.GONE
                            orderModel = result.data
                            //                        list = result.data.getOrderItemDetails();
                            initAdapter()
                        } else {
                            binding.dataLY.visibility = View.GONE
                            binding.noDataLY.noDataLY.visibility = View.VISIBLE
                        }
                    } else {
                        binding.dataLY.visibility = View.GONE
                        binding.noDataLY.noDataLY.visibility = View.GONE
                        binding.failGetDataLY.failGetDataLY.visibility = View.VISIBLE
                        binding.failGetDataLY.failTxt.text = message
                    }
                }            }

        }).getOrderDetails(order_id, user_id, store_id, type)
    }
}